src/common/KokkosKernels_Sorting.hpp

```diff
@@ -0,0 +1,631 @@
+    [=](Ordinal i)
+    {
+      values[i] -= minVal;
+    });
+  }
+  //determine how many significant bits the data has
+  Ordinal sortBits = 0;
+  ValueType upperBound = maxVal - minVal;
+  while(upperBound)
+  {
+    upperBound >>= 1;
+    sortBits++;
+  }
+  Kokkos::single(Kokkos::PerThread(thread),
+  [=]()
+  {
+    //Is the data currently held in values (false) or valuesAux (true)?
+    bool inAux = false;
+    //sort 4 bits at a time, into 16 buckets
+    ValueType mask = 0xF;
+    //maskPos counts the low bit index of mask (0, 4, 8, ...)
+    Ordinal maskPos = 0;
+    for(Ordinal s = 0; s < (sortBits + 3) / 4; s++)
+    {
+      //Count the number of elements in each bucket
+      Ordinal count[16] = {0};
+      Ordinal offset[17];
+      if(!inAux)
+      {
+        for(Ordinal i = 0; i < n; i++)
+        {
+          count[(values[i] & mask) >> maskPos]++;
+        }
+      }
+      else
+      {
+        for(Ordinal i = 0; i < n; i++)
+        {
+          count[(valuesAux[i] & mask) >> maskPos]++;
+        }
+      }
+      offset[0] = 0;
+      //get offset as the prefix sum for count
+      for(Ordinal i = 0; i < 16; i++)
+      {
+        offset[i + 1] = offset[i] + count[i];
+      }
+      //now for each element in [lo, hi), move it to its offset in the other buffer
+      //this branch should be ok because whichBuf is the same on all threads
+      if(!inAux)
+      {
+        //copy from *Over to *Aux
+        for(Ordinal i = 0; i < n; i++)
+        {
+          Ordinal bucket = (values[i] & mask) >> maskPos;
+          valuesAux[offset[bucket + 1] - count[bucket]] = values[i];
+          count[bucket]--;
+        }
+      }
+      else
+      {
+        //copy from *Aux to *Over
+        for(Ordinal i = 0; i < n; i++)
+        {
+          Ordinal bucket = (valuesAux[i] & mask) >> maskPos;
+          values[offset[bucket + 1] - count[bucket]] = valuesAux[i];
+          count[bucket]--;
+        }
+      }
+      inAux = !inAux;
+      mask = mask << 4;
+      maskPos += 4;
+    }
+  });
+  //Move values back into main array if they are currently in aux.
+  //This is the case if an odd number of rounds were done.
+  if(((sortBits + 3) / 4) % 2 == 1)
+  {
+    Kokkos::parallel_for(Kokkos::ThreadVectorRange(thread, n),
+    [=](Ordinal i)
+    {
+      values[i] = valuesAux[i];
+    });
+  }
+  //remove bias to restore original values
+  if(minVal != 0)
+  {
+    Kokkos::parallel_for(Kokkos::ThreadVectorRange(thread, n),
+    [=](Ordinal i)
+    {
+      values[i] += minVal;
+    });
+  }
+}
+
+//Radix sort for integers, on a single thread within a team.
+//While sorting, also permute "perm" array along with the values.
+//Pros: few diverging branches, so good for sorting on a single GPU thread/warp.
+//Con: requires auxiliary storage, this version only works for integers (although float/double is possible)
+template<typename Ordinal, typename ValueType, typename PermType, typename Thread>
```

Comment:
if the above call becomes teamRadixSort, this could be just radixSort

Reply:
This is exactly the same sorting algorithm (thread-local), the only difference is it permutes "perm" along with the values while sorting.